File: iOS/Schema/Sources/ServerDrivenComponent/Screen/NavigationBar.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import UIKit

public struct NavigationBar: Decodable, AutoInitiable {

    public let title: String
    public let styleId: String?
    public let showBackButton: Bool?
    public let backButtonAccessibility: Accessibility?
    public let navigationBarItems: [NavigationBarItem]?

// sourcery:inline:auto:NavigationBar.Init
    public init(
        title: String,
        styleId: String? = nil,
        showBackButton: Bool? = nil,
        backButtonAccessibility: Accessibility? = nil,
        navigationBarItems: [NavigationBarItem]? = nil
    ) {
        self.title = title
        self.styleId = styleId
        self.showBackButton = showBackButton
        self.backButtonAccessibility = backButtonAccessibility
        self.navigationBarItems = navigationBarItems
    }
// sourcery:end
}

public struct NavigationBarItem: Decodable, AccessibilityComponent, IdentifiableComponent {
    
    public let id: String?
    public let image: StringOrExpression?
    public let text: String
    public let action: RawAction
    public let accessibility: Accessibility?

    public init(
        id: String? = nil,
        image: String? = nil,
        text: String,
        action: RawAction,
        accessibility: Accessibility? = nil
    ) {
        self.id = id
        self.image = image
        self.text = text
        self.action = action
        self.accessibility = accessibility
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case image
        case text
        case action
        case accessibility
    }

    enum LocalImageCodingKey: String, CodingKey {
        case mobileId
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        id = try container.decodeIfPresent(String.self, forKey: .id)
        text = try container.decode(String.self, forKey: .text)
        action = try container.decode(forKey: .action)
        accessibility = try container.decodeIfPresent(Accessibility.self, forKey: .accessibility)
        
        let nestedContainer = try? container.nestedContainer(keyedBy: LocalImageCodingKey.self, forKey: .image)
        image = try nestedContainer?.decodeIfPresent(String.self, forKey: .mobileId)
    }

}

File: schema-ruby-erb/model_generation/Models/ServerDriven/Screen/navigation_bar.rb
#   Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA

#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
 
#       http://www.apache.org/licenses/LICENSE-2.0
 
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

require_relative '../../../Synthax/Attributes/variable.rb'
require_relative '../../base_component.rb'
require_relative '../../../Synthax/Types/built_in_type.rb'
require_relative '../server_driven_component.rb'
require_relative '../../Context/context_component.rb'

class NavigationBar < BaseComponent

    def initialize
        variables = [
            Variable.new(
                :name => "title",
                :typeName => TypesToString.string,
                :comment => "define the Title on the navigation bar"
            ),
            Variable.new(
                :name => "styleId",
                :typeName => TypesToString.string,
                :isOptional => true,
                :comment => "could define a custom layout for your action bar/navigation  bar"
            ),
            Variable.new(
                :name => "showBackButton",
                :typeName => TypesToString.bool,
                :isOptional => true,
                :comment => "enable a back button into your action bar/navigation bar"
            ),
            Variable.new(
                :name => "backButtonAccessibility",
                :typeName => Accessibility.new.name,
                :isOptional => true,
                :comment => "define accessibility details for the item"
            ),
            List.new(
                :name => "navigationBarItems",
                :typeName => NavigationBarItem.new.name,
                :isOptional => true,
                :comment => "defines a List of navigation bar items."
            ),
        ]
        synthax_type = BuiltInType.new(
            :name => self.name,
            :variables => variables,
            :package => "br.com.zup.beagle.widget.core",
            :comment => "Typically displayed at the top of the window, containing buttons for navigating within a hierarchy of screens"
        )

        super(synthax_type)

    end

end